### PR TITLE
Update cudnn flash attention

### PR DESCRIPTION
**Pull request by Wong4j**
### PR Category
Others

### PR Types
Others

### Description
cuDNN Flash Attention在Hopper GPU上性能超过开源的Flash Attention。在Ampere GPU上部分case比开源版本慢，目前cuDNN team正在优化中。

PR的改动：
1. cudnn-frontend 从v0.9升级到v1.2.
2. 重构fused_dot_product_attention，增加新的feature，比如bias, GQA/MQA.
3. 重构`fuse_dot_product_attention_pass` IR pass.

解释一下 `incubate/nn/functional/fused_dot_product_attention.py`中添加的两个APIs：
- `cudnn_flash_attention`
  - 这个API具有cuDNN实现的flash attention的完整功能，跟开源版本的flash attention的接口类似。输入参数有`q, k, v, cu_seqlen_q, cu_seqlen_kv`等等，没有`mask`。另外还支持`bias`。
- `fused_dot_product_attention`
  - 其底层调用的也是cuDNN版的flash attention，相当于一个特化版本，这样设计主要是为了在参数和功能上尽量对齐`nn.functional.scaled_dot_product_attention`，输入参数是q, k, v, mask等等。
  - 其中需要注意细节的是，`cudnn_flash_attention`本身没有mask输入，而这个API支持arbitrary mask，内部的实现机制是将`mask`作为`post_scale_bias`传进去。

### CODE REVIEW

**Comment by Wong4j** [comment]
This PR need to test with cudnn>=9.0. Could you help to upgrade cuDNN version in GPUPS. @tianshuo78520a

**Comment by tianshuo78520a** [comment]
> This PR need to test with cudnn>=9.0. Could you help to upgrade cuDNN version in GPUPS. @tianshuo78520a

The GPUPS manager encountered an issue while testing CUDNN9.0 and is still working on resolving it. We need to wait for them to solve the problem before CI can upgrade.

**Comment by Wong4j** [comment]
@tianshuo78520a 有两个问题：
1. 我这个PR有两个UT(test_fused_dot_product_attention_op和test_fused_dot_product_attention_op_static)被添加到了disable_ut中了，所以没有被跑到，导致PR-CI-Coverage报错，在这个里面https://sys-p0.bj.bcebos.com/prec/disable_ut 。请问这个是什么原因？可以移除吗？
2. 我想确认一下。PR-CI-Distribute-stable这个CI，编Paddle的时候开了WITH_CUDNN_FRONTEND吗？

**Comment by tianshuo78520a** [comment]
> 2\. WITH_CUDNN_FRONTEND
第1点，这两个单测好像是会出现随机挂情况，如果解决了可以从列表中移出，暂时没有人力看，我询问下能否豁免Coverage CI。如果有时间，也可以帮看下单测问题。

第2点，你可以在PR-CI-Distribute-stable CI日志中找到，已经是打开状态（WITH_CUDNN_FRONTEND=ON）。
image

**Comment by Wong4j** [comment]
>这两个单测好像是会出现随机挂情况，如果解决了可以从列表中移出


@tianshuo78520a 谢谢回复，这两个单测我本地测是没问题的，能否从列表中移除，跑CI试一下，如果遇到问题我再修

**Comment by tianshuo78520a** [comment]
> > 这两个单测好像是会出现随机挂情况，如果解决了可以从列表中移出
> 
> @tianshuo78520a 谢谢回复，这两个单测我本地测是没问题的，能否从列表中移除，跑CI试一下，如果遇到问题我再修

test_fused_dot_product_attention_op已经解除
test_fused_dot_product_attention_op_static并没有禁用，需要测试你可以提个PR，把单测添加到tools/gpups_test.sh目录下

**Comment by Wong4j** [comment]
@tianshuo78520a 解除后，PR-CI-Distribute-stable中test_fused_dot_product_attention_op这个单测pass了。但PR-CI-Coverage里面仍然显示没有跑到相关代码。
PR-CI-Distribute-stable是不会影响coverage检查的吗？如果是这样的话，就只能麻烦帮忙豁免coverage了

**Comment by tianshuo78520a** [comment]
> @tianshuo78520a 解除后，PR-CI-Distribute-stable中test_fused_dot_product_attention_op这个单测pass了。但PR-CI-Coverage里面仍然显示没有跑到相关代码。 PR-CI-Distribute-stable是不会影响coverage检查的吗？如果是这样的话，就只能麻烦帮忙豁免coverage了

是的Coverage 是V100环境，应该不跑这个单测，我联系负责人豁免下

**Review by tianshuo78520a — APPROVED** [review]

**Comment by Wong4j** [comment]
@onecatcn 这个PR CI都过了，下面这几项麻烦帮忙找对应的人approve。   

PR-CI-APPROVAL
```
0. You must have raindrops2sea or XiaoguangHu01 approval for change 20+ files or add than 1000+ lines of content.
1. You must have one RD (XiaoguangHu01,chenwhql,zhiqiu,Xreki,luotao1,qili93,Aurelius84) approval for the usage of const_cast.
2. Unittest is not allowed to be disabled. You must have one RD (kolinwei(Recommend), wanghuancoder, luotao1, QingshuChen, qili93 or ZzSean or Aurelius84) approval for the usage of @unittest.skip or @unittest.skipIf.
3. The error message you wrote in PADDLE_ENFORCE{_**} or PADDLE_THROW does not meet our error message writing specification. Possible errors include 1. the error message is empty / 2. the error message is too short / 3. the error type is not specified. Please read the specification [ https://github.com/PaddlePaddle/Paddle/wiki/Paddle-Error-Message-Writing-Specification ], then refine the error message. If it is a mismatch, please request chenwhql (Recommend), luotao1 or lanxianghit or Aurelius84 review and approve.
```


PR-CI-Static-Check
```
1. You must have one RD (qingqing01(Recommend), heavengate) approval for the changes of  Inputs/Output/Attrs of OPs. 
2024-05-13 17:15:02  For more details, please click [https://github.com/PaddlePaddle/Paddle/wiki/OP-Input-Output-Attribute-Compatibility-Modification].
  * The added Input 'cu_seqlen_q' is `def`, need inference to review.
  * The added Input 'bias' is `def`, need inference to review.
  * The added Input 'cu_seqlen_kv' is `def`, need inference to review.
  * The added attr 'bias_type_str' is `def`, need inference to review.
  * The added attr 'mask_type_str' is `def`, need inference to review.
2. print or std::cout is not recommended for direct use, please use logging or VLOG. If it is necessary to use, please contact tianshuo78520a (Recommend) or zhangbo9674 review and approve.
```

**Comment by vivienfanghuagood** [comment]
请问新的cudnn_flash_attention，性能上和开源版本flash_attention有差异吗？

**Comment by Wong4j** [comment]
@vivienfanghuagood 目前cuDNN v9.0，在Hopper GPU上是比开源版本快，在Ampere上有一些case慢一些，cuDNN team在优化中

File: python/paddle/incubate/nn/functional/fused_dot_product_attention.py [hunk 1]
```diff
@@ -39,64 +41,76 @@ def fused_dot_product_attention(
...
+
+    # CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT
+    # - unset:       enables workspace optimization when required workspace is <= 256MB
+    #                or when bias gradient needs to be computed
+    # - n:           enables workspace optimization when required workspace is <= n bytes
+    # - -1:          enables workspace optimization always
+    # - 0:           disables workspace optimization always
     if use_workspace_opt is False:
-        os.environ['CUDNN_FUSE_ATTN_USE_WORKSPACE_OPT'] = '0'
+        os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "0"
```
Comment (vivienfanghuagood) [inline python/paddle/incubate/nn/functional/fused_dot_product_attention.py]: 在API里面改变环境变量的值，似乎不太恰当

Comment (Wong4j) [reply python/paddle/incubate/nn/functional/fused_dot_product_attention.py]: 这个是告诉cudnn frontend 开启对应的优化。[cudnn_frontend/node/scaled_dot_product_flash_attention.h#L804](https://github.com/NVIDIA/cudnn-frontend/blob/b740542818f36857acf7f9853f749bbad4118c65/include/cudnn_frontend/node/scaled_dot_product_flash_attention.h#L804)
如果不写在这里，也只能把相同的逻辑移到c++ code里面，有什么建议吗？

File: paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc [hunk 2]
```diff
@@ -118,12 +118,15 @@ class FusedDotProductAttentionPattern : public paddle::drr::DrrPatternBase {
                {{{"scaling_factor", scaling_factor},
```
Comment (yuanlehome) [inline paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc:118]: 是否可以加一下这个pass的单测呢？参考这个单测 test/ir/pir/fused_pass/test_fused_flash_attn_pass.py

Comment (Wong4j) [reply paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc:118]: 对应的单测在之前的PR中已经加上了[test/ir/pir/fused_pass/test_fused_dot_product_attention_pass.py](https://github.com/PaddlePaddle/Paddle/blob/develop/test/ir/pir/fused_pass/test_fused_dot_product_attention_pass.py)。这个PR是在之前的基础上重构了代码，不需要修改这个单测，已经测试过了。

Comment (yuanlehome) [reply paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc:118]: 举个例子，重构之后的接口删去了mask输入，但我看单测里组网还是有mask输入呢

Comment (yuanlehome) [reply paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc:118]: 单测是不是也需要同步修改下？

Comment (Wong4j) [reply paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc:118]: 解释一下，cudnn flash attention本身不能直接输入mask，原来的实现中，OP输入是q,k,v,mask,...，OP内部加了一个手写kernel将mask转成cu_seqlen。而最新的cudnn中支持了bias，OP重构之后输入是q,k,v,bias,...。当bias_type为`post_scale_bias`的时候，bias其实是等效于mask的。所以新的写法是把网络中的mask当做bias传进去。这样的好处是不用再手写kernel将mask转成cu_seqlen，并且还支持arbitrary mask。旧的mask和新的bias同样放在第四个参数，所以单测不用修改。
image

Comment (Wong4j) [reply paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc:118]: arbitrary mask的解释：
左边是普通的padding mask，只有尾部补0。右边是arbitrary mask，任意部位都可以是0，没有限制。
[1,1,1,0,0]     [0,1,0,1,0]
[1,1,1,0,0]     [1,0,1,0,0]
[1,1,1,0,0]     [0,1,0,0,1]
[0,0,0,0,0]     [0,0,1,0,1]

File: python/paddle/incubate/nn/functional/fused_dot_product_attention.py (hunk 1 above)
Comment (Wong4j) [reply python/paddle/incubate/nn/functional/fused_dot_product_attention.py]: 这个功能必须要setenv才能开启，放在C++ code里面可以吗？还有其他的review意见吗？

File: paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc (hunk 2 above)
Comment (Wong4j) [reply paddle/fluid/pir/transforms/gpu/fused_dot_product_attention_pass.cc:118]: 还有其他的review意见吗？

File: python/paddle/incubate/nn/functional/fused_dot_product_attention.py (hunk 1 above)
Comment (Wong4j) [reply python/paddle/incubate/nn/functional/fused_dot_product_attention.py]: 已经在API里面去掉了环境变量的修改

**Review by yuanlehome — DISMISSED** [review]

**Comment by Wong4j** [comment]
@vivienfanghuagood 请问还有review意见吗？

**Comment by Wong4j** [comment]
review意见都修改了
@qingqing01 @kolinwei @chenwhql 麻烦看一下是否可以approve

**Review by vivienfanghuagood — APPROVED** [review]

**Comment by Wong4j** [comment]
@onecatcn 麻烦帮忙提醒一下这几位

> review意见都修改了 @qingqing01 @kolinwei @chenwhql 麻烦看一下是否可以approve

**Review by kolinwei — DISMISSED** [review]

File: python/paddle/incubate/nn/functional/fused_dot_product_attention.py [hunk 3]
```diff
@@ -105,17 +105,132 @@ def fused_dot_product_attention(
...
+                'out': [out],
+                'softmax_out': [softmax_out],
+                'rng_state': [rng_state],
+            },
+            attrs=attrs,
+        )
+        return out
+
+
+def fused_dot_product_attention(
```
Comment (qingqing01) [inline python/paddle/incubate/nn/functional/fused_dot_product_attention.py:132]: API的角度， https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/functional/flash_attention.py#L722C5-L722C33 这里已有 scaled_dot_product_attention ，需说明和这个API的差异、推荐使用场景。

**Review by qingqing01 — COMMENTED** [review]
此外看到PASS似乎实现了，但没开启？

File: python/paddle/incubate/nn/functional/fused_dot_product_attention.py (hunk 3 above)
Comment (Wong4j) [reply python/paddle/incubate/nn/functional/fused_dot_product_attention.py:132]: Done. 
在API的注释中说明了和 scaled_dot_product_attention的区别，并推荐在最新的GPU架构中使用。
https://github.com/PaddlePaddle/Paddle/blob/ac9b059e0572e2614c5bb600cd05894f0cbdd7b6/python/paddle/incubate/nn/functional/fused_dot_product_attention.py#L162-L169

**Review by qingqing01 — APPROVED** [review]

**Review by Aurelius84 — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]
